### PR TITLE
Record and use slice index for stacked netdcf + usability fixes

### LINKED ISSUE #415: Mapping from `Dataset` to `Storage Unit` is poorly defined when using netcdf storage
When using netcdf files for storage we often "stack" year worth of observations into one netcdf file. As a result multiple `dc.Dataset`s point to the same file on disk. Since `dc.Dataset` represents a single point in time only one slice along the time dimension "belongs" to a given dataset. Which slice that is, is poorly defined.

We do not store time slice index anywhere in the `dc.Dataset` metadata document. Instead `center_time` property is used to "locate" the index that "matches best". 

The "matches best" logic is fairly involved, see:

https://github.com/opendatacube/datacube-core/blob/5f7211d2eb26ac10d6dfb754418d4bd3079cdf7a/datacube/storage/storage.py#L473-L501

It can not use exact equality since `center_time` is a `datetime.datetime`, which can contain sub-second component, while what's stored on disk is `float64` containing seconds since 1970, and is always rounded to seconds (I only observed it being rounded down, but who knows, it happens inside netcdf library). By the way this process is repeated for every band being loaded, even though they all share the same time slice index.

It means that file needs to be opened before one can figure out what part of the file is to be read. This is a problem for optimized read scheduling.

### PR DESCRIPTION
**Pull request by Kirill888**
### Reason for this pull request

- Usability fixes for `datacube dataset add/update`
- Better handling of multi-tenanted files (i.e. stacked netcdf)

### Proposed changes

1. Remove product matching rules options from `datacube dataset update`, it operates on existing datasets only, it should need to auto-match anything.
2. Remove `--match-rules` option from `datacube dataset add`, this one wasn't being tested, nor was it documented. 
3. Deprecate `--auto-match` option, it was on by default anyway.
4. Rename `--dtype` to `--product` (discussed in #450)
5. `read_documents` can report url that includes part of the multi-tenant file form which documet was read
6. Indexing/update process records time slice index as part of the location URI, when working with stacked netcdf files
7. Stacker records time slice index as part of the location URI
8. When loading data use recorded index instead of searching for best matching timestamp

### Incidental changes

- Enable compression of metadata text during ingestion #452 
- Fix chunking configuration for strings in netcdf files  #453 

------------------------------------------------------------------------
 - [x] Closes #415 #447 #448 #452 #453 #454 #445
 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

File: datacube/utils/__init__.py [hunk 1]
```diff
@@ -238,7 +291,28 @@ def remove_implicit_resolver(cls, tag_to_remove):
...
+    """
+    lineage = doc.get('lineage', None)
+    if lineage is None:
+        return doc
+
+    if 'source_datasets' not in lineage:
+        return doc
+
+    if inplace:
+        doc['lineage']['source_datasets'] = {}
```
Comment (omad) [inline datacube/utils/__init__.py]: Unfortunately, this (and the rest) of dataset documents is configurable with the MetadataType thing. See https://github.com/opendatacube/datacube-core/blob/fbefaa5f95704160133cd45df4cd5a12e1ea88f9/datacube/index/default-metadata-types.yaml#L54 for where it is defined for the `eo` metadata type. (Lodge complaints with @jeremyh. :) I can see the appeal of configurability, but I think other metadata standards will be *more* different than what we can simply configure away, and we pay a significant cost in complexity...)

Fortunately, in the two places this function is used it does have access to the Product, which is of a particular MetadataType, so can lookup what offset `sources` will be at. Still a total pain though...

**Review by omad — CHANGES_REQUESTED** [review]
All looks amazing, except for the annoyingness with dataset document structure.

Can you please also copy the proposed and incidental changes into the `whats_new` doc.

File: datacube/utils/__init__.py (hunk 1 above)
Comment (Kirill888) [reply datacube/utils/__init__.py]: Good point, I keep forgetting about this indirection. It does mean that `sources` can not be skipped until top-level product is resolved (auto-matched), not a problem in `update`, but we do need this in `add` as well. I'll have a look what that implies.

**Review by omad — APPROVED** [review]
Thank you! 👍

**Pull request merged**